pkg/core/blockchain.go

```diff
@@ -353,6 +353,9 @@ func (bc *Blockchain) init() error {
 	bc.dao.Version = ver
 	bc.persistent.Version = ver
 
+	// Always try to remove garbage. If there is nothing to do, it will exit quickly.
+	go bc.removeOldStorageItems()
```

Comment:
It may safely be done inside `jumpToStateInternal`, but for blockchain initialisation we need more complicated logic. Consider the situation when state sync process was started, and some of the storage items (but not all) were fetched, and the process was interrupted after that. Currently, we're able to recover from that point after next node resumption, i.e. if the chain is not too hight we will continue missing storage items collection. So if we run `go bc.removeOldStorageItems()`, then it will remove all items we've collected, so state sync process is broken by this line.

I suggest the following solution: inside the `jumpToStateInternal(...)` method we should store some flag in the DB right before the goroutine is started; this flag means that removal was started. And at the end of `removeOldStorageItems()` we may remove this flag from the DB, meaning that the removal process was ended. Then during blockchain initialisation we should start this goroutine iff the flag is still in the storage, i.e. not all data were removed.

Reply:
Will just checking it `jumpStage >= newStorageItemsAdded` do the trick?

Reply:
Yes, nice solution.